File: src/mission/uav_mission/uav_mission/mock_mission.py
"""MockMission: publishes a fixed, latched Mission on startup and reports
PlannerStatus, so the rest of the pipeline can be exercised without a real
mission file loader or operator input UI.
"""
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy

from uav_interfaces.msg import Mission, Waypoint, PlannerStatus


class MockMission(Node):
    def __init__(self):
        super().__init__('mock_mission')
        mission_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST, depth=1)
        sensor_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            history=HistoryPolicy.KEEP_LAST, depth=5)

        self.mission_pub = self.create_publisher(Mission, '/mission/current', mission_qos)
        self.create_subscription(PlannerStatus, '/planning/status', self.on_status, sensor_qos)

        self.publish_mission()
        self.timer = self.create_timer(1.0, self.publish_mission)

    def publish_mission(self):
        m = Mission()
        m.header.stamp = self.get_clock().now().to_msg()
        m.header.frame_id = 'map'
        m.mission_id = 'golden-scenario-01'

        wp = Waypoint()
        wp.position.x = 10.0
        wp.position.y = 0.0
        wp.position.z = 3.0
        wp.acceptance_radius = 0.5
        wp.waypoint_type = Waypoint.TYPE_TARGET
        m.waypoints = [wp]

        m.max_speed = 3.0
        m.boundary_radius = 50.0
        m.min_altitude = 1.0
        m.max_altitude = 20.0
        self.mission_pub.publish(m)

    def on_status(self, msg: PlannerStatus):
        self.get_logger().debug(f'planner state={msg.state} progress={msg.progress}')


def main(args=None):
    rclpy.init(args=args)
    node = MockMission()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/mission/uav_mission/src/mock_mission.cpp
// MockMission: publishes a fixed, latched Mission on startup and logs
// PlannerStatus, so the rest of the pipeline can be exercised without a
// real mission file loader or operator input UI.
#include <chrono>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "uav_interfaces/msg/mission.hpp"
#include "uav_interfaces/msg/waypoint.hpp"
#include "uav_interfaces/msg/planner_status.hpp"

using namespace std::chrono_literals;
using uav_interfaces::msg::Mission;
using uav_interfaces::msg::Waypoint;
using uav_interfaces::msg::PlannerStatus;

class MockMission : public rclcpp::Node
{
public:
  MockMission() : Node("mock_mission")
  {
    rclcpp::QoS mission_qos(rclcpp::KeepLast(1));
    mission_qos.reliable().transient_local();
    rclcpp::QoS sensor_qos(rclcpp::KeepLast(5));
    sensor_qos.best_effort();

    mission_pub_ = create_publisher<Mission>("/mission/current", mission_qos);
    status_sub_ = create_subscription<PlannerStatus>(
      "/planning/status", sensor_qos,
      [this](const PlannerStatus::SharedPtr msg) {
        RCLCPP_DEBUG(
          get_logger(), "planner state=%u progress=%.2f", msg->state, msg->progress);
      });

    publish_mission();
    timer_ = create_wall_timer(1s, std::bind(&MockMission::publish_mission, this));
  }

private:
  void publish_mission()
  {
    Mission m;
    m.header.stamp = this->now();
    m.header.frame_id = "map";
    m.mission_id = "golden-scenario-01";

    Waypoint wp;
    wp.position.x = 10.0;
    wp.position.y = 0.0;
    wp.position.z = 3.0;
    wp.acceptance_radius = 0.5F;
    wp.waypoint_type = Waypoint::TYPE_TARGET;
    m.waypoints.push_back(wp);

    m.max_speed = 3.0F;
    m.boundary_radius = 50.0F;
    m.min_altitude = 1.0F;
    m.max_altitude = 20.0F;
    mission_pub_->publish(m);
  }

  rclcpp::Publisher<Mission>::SharedPtr mission_pub_;
  rclcpp::Subscription<PlannerStatus>::SharedPtr status_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MockMission>());
  rclcpp::shutdown();
  return 0;
}
